File: web/src/store/personas.ts
import { create } from "zustand"
import { api } from "@/lib/api"

export interface Persona {
    id: string
    name: string
    description: string
    system_prompt: string
    icon: string
    color: string
    allowed_tools: string[]
    is_builtin: boolean
    created_at: string
    updated_at: string
}

interface PersonaState {
    personas: Persona[]
    activePersonaId: string | null
    isLoading: boolean

    // Actions
    fetchPersonas: () => Promise<void>
    setActivePersona: (id: string | null) => void
    createPersona: (data: {
        name: string
        system_prompt: string
        description?: string
        icon?: string
        color?: string
        allowed_tools?: string[]
    }) => Promise<Persona | null>
    updatePersona: (
        id: string,
        data: {
            name?: string
            description?: string
            system_prompt?: string
            icon?: string
            color?: string
            allowed_tools?: string[]
        }
    ) => Promise<void>
    deletePersona: (id: string) => Promise<void>
}

export const usePersonaStore = create<PersonaState>((set, get) => ({
    personas: [],
    activePersonaId: null,
    isLoading: false,

    fetchPersonas: async () => {
        set({ isLoading: true })
        try {
            const { data } = await api.GET("/v1/personas")
            if (data) {
                const personas = (data as unknown as Persona[]) ?? []
                set({ personas })
            }
        } finally {
            set({ isLoading: false })
        }
    },

    setActivePersona: (id) => set({ activePersonaId: id }),

    createPersona: async (body) => {
        const { data } = await api.POST("/v1/personas", { body })
        if (data) {
            const persona = data as unknown as Persona
            set((s) => ({ personas: [...s.personas, persona] }))
            return persona
        }
        return null
    },

    updatePersona: async (id, body) => {
        const { data } = await api.PATCH("/v1/personas/{id}", {
            params: { path: { id } },
            body,
        })
        if (data) {
            const updated = data as unknown as Persona
            set((s) => ({
                personas: s.personas.map((p) => (p.id === id ? updated : p)),
            }))
        }
    },

    deletePersona: async (id) => {
        await api.DELETE("/v1/personas/{id}", {
            params: { path: { id } },
        })
        const state = get()
        set({
            personas: state.personas.filter((p) => p.id !== id),
            ...(state.activePersonaId === id ? { activePersonaId: null } : {}),
        })
    },
}))

File: pkg/kernel/personas.go
package kernel

import (
	"context"
	"time"

	"github.com/manthysbr/auleOS/internal/core/domain"
)

// domainPersonaToAPI converts a domain Persona to an API Persona
func domainPersonaToAPI(p domain.Persona) Persona {
	id := string(p.ID)
	name := p.Name
	desc := p.Description
	prompt := p.SystemPrompt
	icon := p.Icon
	color := p.Color
	builtin := p.IsBuiltin
	createdAt := p.CreatedAt
	updatedAt := p.UpdatedAt

	var allowed *[]string
	if len(p.AllowedTools) > 0 {
		allowed = &p.AllowedTools
	}

	var modelOverride *string
	if p.ModelOverride != "" {
		mo := p.ModelOverride
		modelOverride = &mo
	}

	return Persona{
		Id:            &id,
		Name:          &name,
		Description:   &desc,
		SystemPrompt:  &prompt,
		Icon:          &icon,
		Color:         &color,
		AllowedTools:  allowed,
		ModelOverride: modelOverride,
		IsBuiltin:     &builtin,
		CreatedAt:     &createdAt,
		UpdatedAt:     &updatedAt,
	}
}

// ListPersonas implements StrictServerInterface
func (s *Server) ListPersonas(ctx context.Context, _ ListPersonasRequestObject) (ListPersonasResponseObject, error) {
	personas, err := s.repo.ListPersonas(ctx)
	if err != nil {
		s.logger.Error("failed to list personas", "error", err)
		return ListPersonas200JSONResponse([]Persona{}), nil
	}

	result := make([]Persona, 0, len(personas))
	for _, p := range personas {
		result = append(result, domainPersonaToAPI(p))
	}
	return ListPersonas200JSONResponse(result), nil
}

// CreatePersona implements StrictServerInterface
func (s *Server) CreatePersona(ctx context.Context, request CreatePersonaRequestObject) (CreatePersonaResponseObject, error) {
	now := time.Now()
	p := domain.Persona{
		ID:           domain.NewPersonaID(),
		Name:         request.Body.Name,
		SystemPrompt: request.Body.SystemPrompt,
		IsBuiltin:    false,
		CreatedAt:    now,
		UpdatedAt:    now,
	}

	if request.Body.Description != nil {
		p.Description = *request.Body.Description
	}
	if request.Body.Icon != nil {
		p.Icon = *request.Body.Icon
	} else {
		p.Icon = "bot"
	}
	if request.Body.Color != nil {
		p.Color = *request.Body.Color
	} else {
		p.Color = "blue"
	}
	if request.Body.AllowedTools != nil {
		p.AllowedTools = *request.Body.AllowedTools
	}
	if request.Body.ModelOverride != nil {
		p.ModelOverride = *request.Body.ModelOverride
	}

	if err := s.repo.CreatePersona(ctx, p); err != nil {
		s.logger.Error("failed to create persona", "error", err)
		return CreatePersona201JSONResponse(domainPersonaToAPI(p)), nil
	}

	return CreatePersona201JSONResponse(domainPersonaToAPI(p)), nil
}

// GetPersona implements StrictServerInterface
func (s *Server) GetPersona(ctx context.Context, request GetPersonaRequestObject) (GetPersonaResponseObject, error) {
	p, err := s.repo.GetPersona(ctx, domain.PersonaID(request.Id))
	if err != nil {
		if err == domain.ErrPersonaNotFound {
			return GetPersona404JSONResponse{Error: strPtr("persona not found")}, nil
		}
		s.logger.Error("failed to get persona", "error", err)
		return GetPersona404JSONResponse{Error: strPtr(err.Error())}, nil
	}
	return GetPersona200JSONResponse(domainPersonaToAPI(p)), nil
}

// UpdatePersona implements StrictServerInterface
func (s *Server) UpdatePersona(ctx context.Context, request UpdatePersonaRequestObject) (UpdatePersonaResponseObject, error) {
	existing, err := s.repo.GetPersona(ctx, domain.PersonaID(request.Id))
	if err != nil {
		if err == domain.ErrPersonaNotFound {
			return UpdatePersona404JSONResponse{Error: strPtr("persona not found")}, nil
		}
		return UpdatePersona404JSONResponse{Error: strPtr(err.Error())}, nil
	}

	if request.Body.Name != nil {
		existing.Name = *request.Body.Name
	}
	if request.Body.Description != nil {
		existing.Description = *request.Body.Description
	}
	if request.Body.SystemPrompt != nil {
		existing.SystemPrompt = *request.Body.SystemPrompt
	}
	if request.Body.Icon != nil {
		existing.Icon = *request.Body.Icon
	}
	if request.Body.Color != nil {
		existing.Color = *request.Body.Color
	}
	if request.Body.AllowedTools != nil {
		existing.AllowedTools = *request.Body.AllowedTools
	}
	if request.Body.ModelOverride != nil {
		existing.ModelOverride = *request.Body.ModelOverride
	}
	existing.UpdatedAt = time.Now()

	if err := s.repo.UpdatePersona(ctx, existing); err != nil {
		s.logger.Error("failed to update persona", "error", err)
		return UpdatePersona404JSONResponse{Error: strPtr(err.Error())}, nil
	}

	return UpdatePersona200JSONResponse(domainPersonaToAPI(existing)), nil
}

// DeletePersona implements StrictServerInterface
func (s *Server) DeletePersona(ctx context.Context, request DeletePersonaRequestObject) (DeletePersonaResponseObject, error) {
	// Prevent deleting builtin personas
	p, err := s.repo.GetPersona(ctx, domain.PersonaID(request.Id))
	if err != nil {
		if err == domain.ErrPersonaNotFound {
			return DeletePersona404Response{}, nil
		}
		return DeletePersona404Response{}, nil
	}
	if p.IsBuiltin {
		return DeletePersona404Response{}, nil // refuse silently
	}

	if err := s.repo.DeletePersona(ctx, domain.PersonaID(request.Id)); err != nil {
		return DeletePersona404Response{}, nil
	}
	return DeletePersona204Response{}, nil
}

// strPtr is a helper to create a *string from a string literal
func strPtr(s string) *string {
	return &s
}
